runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/environment/DockerWrapper.java

```diff
@@ -0,0 +1,125 @@
+  }
+
+  private final String dockerExecutable;
+  private final Duration commandTimeout;
+
+  private DockerWrapper(String dockerExecutable, Duration commandTimeout) {
+    this.dockerExecutable = dockerExecutable;
+    this.commandTimeout = commandTimeout;
+  }
+
+  /**
+   * Runs the given container image with the given command line arguments. Returns the running
+   * container id.
+   */
+  public String runImage(String imageTag, List<String> args) throws IOException, TimeoutException,
+      InterruptedException {
+    checkArgument(!imageTag.isEmpty(), "Docker image tag required");
+    // TODO: Validate args?
+    return runShortCommand(ImmutableList.<String>builder()
+        .add(dockerExecutable)
+        .add("run")
+        .add("-d")
+        .add(imageTag)
+        .addAll(args)
+        .build());
+  }
+
+  /**
+   * Kills a docker container by container id.
+   * @throws IOException if an IOException occurs or if the given container id does not exist
+   */
+  public void killContainer(String containerId) throws IOException, TimeoutException,
+      InterruptedException {
+    checkNotNull(containerId);
+    checkArgument(CONTAINER_ID_PATTERN.matcher(containerId).matches(),
+        "Container ID must be a 64-character hexadecimal string");
+    runShortCommand(Arrays.asList(dockerExecutable, "kill", containerId));
+  }
+
+  /** Run the given command invocation and return stdout as a String. */
+  private String runShortCommand(List<String> invocation) throws IOException,
+      TimeoutException, InterruptedException {
+    ProcessBuilder pb = new ProcessBuilder(invocation);
+    Process process = pb.start();
+    // TODO: Consider supplying executor service here.
+    CompletableFuture<String> resultString = CompletableFuture.supplyAsync(() -> {
+      // NOTE: We do not own the underlying stream and do not close it.
+      BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream(),
+          StandardCharsets.UTF_8));
+      return reader.lines().collect(Collectors.joining());
+    });
+    // TODO: Retry on interrupt?
+    boolean processDone = process.waitFor(commandTimeout.toMillis(), TimeUnit.MILLISECONDS);
+    if (!processDone) {
+      process.destroy();
+      throw new TimeoutException(
+          String.format("Timed out while waiting for command '%s'",
+              invocation.stream().collect(Collectors.joining(" "))));
+    }
+    int exitCode = process.exitValue();
+    if (exitCode != 0) {
+      throw new IOException(String.format("Received exit code %d for command '%s'",
+          exitCode, invocation.stream().collect(Collectors.joining(" "))));
+    }
+    try {
+      // TODO: Consider a stricter timeout.
+      return resultString.get(commandTimeout.toMillis(), TimeUnit.MILLISECONDS);
+    } catch (ExecutionException e) {
+      Throwable cause = e.getCause();
+      // Recast any exceptions in reading output as IOExceptions.
+      if (cause instanceof IOException) {
```

Comment:
This style of unwrapping is moderately discouraged because it hides the separate-thread nature of the call - consider just rewrapping in all cases.